File: internal/icu/units.go
package icu

import (
	"sort"
	"strings"
	"sync"
)

// How a language writes a measurement.
//
// "987 km/h" is "987 km/h" in German too but "987公里/小時" in Chinese, and
// "1 hour" is "1 Stunde" and "2 hours" is "2 Stunden": the name of the unit,
// whether a space goes before it, which form it takes with one and with many,
// and how a unit written underneath another is joined to it. None of that can
// be worked out from the name, so all of it is carried.
//
// It is kept apart from the rest of the locale data and unpacked only when a
// program asks for a measurement, since most never do.

// The units a number may be written in.
//
// Not every unit in the world: the ones ECMA-402 sanctions, which is a list of
// forty-five chosen because every language has a name for each of them. A unit
// may also be one of these divided by another -- kilometre per hour, litre per
// hundred kilometres -- which is why they are a set rather than a table.
var sanctionedUnits = map[string]bool{
	"acre": true, "bit": true, "byte": true, "celsius": true,
	"centimeter": true, "day": true, "degree": true, "fahrenheit": true,
	"fluid-ounce": true, "foot": true, "gallon": true, "gigabit": true,
	"gigabyte": true, "gram": true, "hectare": true, "hour": true,
	"inch": true, "kilobit": true, "kilobyte": true, "kilogram": true,
	"kilometer": true, "liter": true, "megabit": true, "megabyte": true,
	"meter": true, "microsecond": true, "mile": true, "mile-scandinavian": true,
	"milliliter": true, "millimeter": true, "millisecond": true, "minute": true,
	"month": true, "nanosecond": true, "ounce": true, "percent": true,
	"petabyte": true, "pound": true, "second": true, "stone": true,
	"terabit": true, "terabyte": true, "week": true, "yard": true, "year": true,
}

// HasUnit reports whether a unit is one a number may be written in.
func HasUnit(name string) bool { return sanctionedUnits[name] }

// Units lists them, in order.
func Units() []string {
	out := make([]string, 0, len(sanctionedUnits))
	for name := range sanctionedUnits {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

var (
	unitsOnce           sync.Once
	unitBlocks          map[string]map[string]string
	unitsByTag          map[string]string
	durationUnsupported map[string]bool
	unitFormSep         = ""
)

func loadUnits() {
	unitsOnce.Do(func() {
		text, err := inflate(unitsPacked)
		if err != nil {
			return
		}
		unitBlocks = map[string]map[string]string{}
		unitsByTag = map[string]string{}
		blocks, rest, _ := strings.Cut(text, "\n\n")
		tags, unsupported, _ := strings.Cut(rest, "\n\n")
		var current map[string]string
		for _, line := range strings.Split(blocks, "\n") {
			if line == "" {
				continue
			}
			key, value, ok := strings.Cut(line, "\t")
			if !ok {
				// A line without a tab names the block that follows.
				current = map[string]string{}
				unitBlocks[key] = current
				continue
			}
			if current != nil {
				current[key] = value
			}
		}
		for _, line := range strings.Split(tags, "\n") {
			tag, at, ok := strings.Cut(line, "\t")
			if ok {
				unitsByTag[tag] = at
			}
		}
		durationUnsupported = map[string]bool{}
		for _, tag := range strings.Split(unsupported, "\n") {
			if tag != "" {
				durationUnsupported[tag] = true
			}
		}
	})
}

// HasDurationLocale reports whether DurationFormat carries data for a locale.
// Other Intl constructors support a few languages that DurationFormat does
// not, so their requests must be allowed to fall through to another locale.
func HasDurationLocale(tag string) bool {
	loadUnits()
	if !Has(tag) {
		return false
	}
	resolved := ResolveTag(tag)
	return !durationUnsupported[resolved]
}

// UnitPattern is how this language writes a count of a unit, with {0} where
// the number goes: "{0} km", "{0} Stunden". The form is the plural category
// the count falls in; a language that has no such form answers with the one it
// uses for everything else.
func (l *Locale) UnitPattern(unit, width, category string) (string, bool) {
	forms, ok := l.unitForms(width + "/" + unit)
	if !ok {
		return "", false
	}
	for _, form := range strings.Split(forms, unitFormSep) {
		name, pattern, ok := strings.Cut(form, "=")
		if ok && name == category {
			return pattern, true
		}
	}
	// A language that has no such form is answered with the one it uses for
	// everything else, or failing that with whichever it has.
	all := strings.Split(forms, unitFormSep)
	for _, form := range all {
		name, pattern, ok := strings.Cut(form, "=")
		if ok && name == "other" {
			return pattern, true
		}
	}
	if len(all) > 0 {
		if _, pattern, ok := strings.Cut(all[0], "="); ok {
			return pattern, true
		}
	}
	return "", false
}

// UnitPer is a unit written underneath another, with {0} where the one above
// goes: "{0}/h" makes kilometres per hour out of kilometres.
func (l *Locale) UnitPer(unit, width string) (string, bool) {
	return l.unitForms(width + "!" + unit)
}

// UnitCompound is a pair a language writes its own way rather than by joining
// the two: Japanese says "km/h" where joining would give "km/時間".
func (l *Locale) UnitCompound(unit, width string) (string, bool) {
	return l.unitForms(width + "?" + unit)
}

// DurationListPattern returns the separators used to join duration units.
// DurationFormat supports locales that ListFormat does not, so these patterns
// are carried with the unit data rather than inferred from ListFormat.
func (l *Locale) DurationListPattern(style string) (ListPattern, bool) {
	value, ok := l.unitForms("duration-list/" + style)
	if !ok {
		return ListPattern{}, false
	}
	parts := strings.Split(value, "\x02")
	if len(parts) != 4 {
		return ListPattern{}, false
	}
	return ListPattern{
		Pair:  "{0}" + parts[0] + "{1}",
		Start: parts[1], Middle: parts[2], End: parts[3],
	}, true
}

func (l *Locale) unitForms(key string) (string, bool) {
	loadUnits()
	tag := l.Tag
	if at, ok := unitsByTag[tag]; ok {
		tag = at
	}
	block, ok := unitBlocks[tag]
	if !ok {
		block, ok = unitBlocks["en"]
		if !ok {
			return "", false
		}
	}
	value, ok := block[key]
	return value, ok
}

File: internal/icu/internal/cldrgen/units.mjs
// units.mjs reads how a language writes a measurement.
//
// Usage:
//
//	node internal/icu/internal/cldrgen/units.mjs > .../units.json
//
// "987 km/h" in English is "987 km/h" in German and "987公里/小時" in Chinese:
// the name of the unit, whether a space goes before it, which form it takes
// with one and with many, and how a unit written underneath another is joined
// to it. All of that is per language, and none of it can be worked out from
// the name of the unit.
//
// Forty-five units are carried, which are the ones ECMA-402 sanctions, in the
// three widths it offers. Locales that write them all the same way -- and many
// do, since a regional variant rarely renames a kilometre -- share one entry.

import fs from "fs";
import path from "path";
import {fileURLToPath} from "url";

const here = path.dirname(fileURLToPath(import.meta.url));
const {locales} = JSON.parse(fs.readFileSync(path.join(here, "locales.json"), "utf8"));

const UNITS = ["acre", "bit", "byte", "celsius", "centimeter", "day", "degree",
  "fahrenheit", "fluid-ounce", "foot", "gallon", "gigabit", "gigabyte", "gram",
  "hectare", "hour", "inch", "kilobit", "kilobyte", "kilogram", "kilometer",
  "liter", "megabit", "megabyte", "meter", "microsecond", "mile",
  "mile-scandinavian", "milliliter", "millimeter", "millisecond", "minute",
  "month", "nanosecond", "ounce", "percent", "petabyte", "pound", "second",
  "stone", "terabit", "terabyte", "week", "yard", "year"];

// What stands between one form and the next, which cannot be a character a
// unit is written with.
const SEP = "\u0001";
const LIST_SEP = "\u0002";

// Find one count for every plural form rather than relying on a short list of
// common cases. Cebuano, for example, uses "one" for 0-3 but "other" for 4,
// while several Romance languages only expose "many" at a million.
function pluralSamples(plural) {
  const wanted = new Set(plural.resolvedOptions().pluralCategories);
  const found = new Map();
  const consider = n => {
    const category = plural.select(n);
    if (!found.has(category)) found.set(category, n);
  };

  for (let n = 0; n < 200 && found.size < wanted.size; n++) consider(n);
  for (let whole = 0; whole < 20 && found.size < wanted.size; whole++) {
    for (const fraction of [0.1, 0.2, 0.5, 0.01, 0.02, 0.05]) {
      consider(whole + fraction);
    }
  }
  for (let power = 2; power <= 9 && found.size < wanted.size; power++) {
    const base = 10 ** power;
    for (const delta of [-2, -1, 0, 1, 2]) consider(base + delta);
  }

  const missing = [...wanted].filter(category => !found.has(category));
  if (missing.length !== 0) {
    throw new Error("no unit sample for plural categories: " + missing.join(", "));
  }
  return [...found].map(([category, n]) => ({category, n}));
}

function unitsFor(locale) {
  const out = {};
  const plural = new Intl.PluralRules(locale);
  const samples = pluralSamples(plural);
  const plain = new Intl.NumberFormat(locale);
  for (const width of ["short", "narrow", "long"]) {
    for (const unit of UNITS) {
      const f = new Intl.NumberFormat(locale, {style: "unit", unit, unitDisplay: width});
      const forms = [];
      for (const {category: key, n} of samples) {
        const text = f.format(n);
        const shown = plain.format(n);
        const at = text.indexOf(shown);
        forms.push(key + "=" +
          (at < 0 ? text : text.slice(0, at) + "{0}" + text.slice(at + shown.length)));
      }
      out[width + "/" + unit] = forms.join(SEP);
    }
    // The pairs a language writes its own way rather than by joining the two:
    // Japanese says "km/h" where joining would give "km/時間", and says it in
    // front of the number in the long form.
    const numerators = {};
    for (const u of UNITS) {
      numerators[u] = new Intl.NumberFormat(locale,
        {style: "unit", unit: u, unitDisplay: width}).format(1);
    }

    // What a unit looks like written underneath another: "km/h" is the
    // kilometre with the hour's form after it.
    const numerator = new Intl.NumberFormat(locale,
      {style: "unit", unit: "meter", unitDisplay: width}).format(1);
    for (const unit of UNITS) {
      const both = new Intl.NumberFormat(locale,
        {style: "unit", unit: "meter-per-" + unit, unitDisplay: width}).format(1);
      // Where the numerator stood, since a language may put the denominator
      // in front of it rather than after: Chinese writes "per hour, 1 metre".
      const at = both.indexOf(numerator);
      out[width + "!" + unit] = at < 0 ? both
        : both.slice(0, at) + "{0}" + both.slice(at + numerator.length);
    }
    for (const above of UNITS) {
      for (const below of UNITS) {
        if (above === below) continue;
        const actual = new Intl.NumberFormat(locale,
          {style: "unit", unit: above + "-per-" + below, unitDisplay: width}).format(1);
        const joined = out[width + "!" + below].replace("{0}", numerators[above]);
        if (actual === joined) continue;
        const at = actual.indexOf(numerators[above]);
        if (at < 0) continue;
        out[width + "?" + above + "-per-" + below] =
          actual.slice(0, at) + "{0}" + actual.slice(at + numerators[above].length);
      }
    }
  }

  for (const style of ["long", "short", "narrow"]) {
    const formatter = new Intl.DurationFormat(locale, {style});
    const separators = value => formatter.formatToParts(value)
      .filter(part => part.type === "literal" && part.unit === undefined)
      .map(part => part.value);
    const pair = separators({years: 3, months: 4});
    const many = separators({years: 3, months: 4, weeks: 5, days: 6});
    if (pair.length === 1 && many.length === 3) {
      out["duration-list/" + style] = [pair[0], ...many].join(LIST_SEP);
    }
  }
  return out;
}

// The locales that say everything the same way share one entry.
const blocks = new Map();
const same = {};
for (const locale of locales) {
  let block;
  try { block = unitsFor(locale); } catch (e) { continue; }
  const key = JSON.stringify(block);
  if (!blocks.has(key)) blocks.set(key, {at: locale, block});
  same[locale] = blocks.get(key).at;
}
const units = {};
for (const {at, block} of blocks.values()) units[at] = block;

// DurationFormat is newer than NumberFormat's unit support, and ICU does not
// necessarily carry it for every locale supported elsewhere. Keep that
// distinction so an unsupported request falls through to the next locale.
const durationUnsupported = locales.filter(locale =>
  Intl.DurationFormat.supportedLocalesOf(locale).length === 0);

process.stdout.write(JSON.stringify({
  icu: process.versions.icu, units, same, durationUnsupported,
}) + "\n");
